### PR TITLE
Fixes #31680 - execute service actions in group

### PR DESCRIPTION
**Pull request by upadhyeammit**
This handles service actions in group. The services are grouped in priority and with this change with execute an action on same priority services at a time. This improves the start and stop times,
Existing sequential handling of service actions,
~~~
# time bin/foreman-maintain service restart
Running Restart Services

real	3m18.994s
user	0m2.560s
sys	0m1.176s
~~~

Improved group handling of service actions,
~~~
# time bin/foreman-maintain service restart
Running Restart Services

real	2m23.988s
user	0m2.465s
sys	0m0.997s
~~~

### CODE REVIEW

**Comment by upadhyeammit** [comment]
@Gauravtalreja1 request to review and test!

**Comment by mccun934** [comment]
Tested this and replicated the above results, shaved off a minute:

before patch:
```
real	3m11.152s
user	0m1.923s
sys	0m0.813s

```

after:

```
real	2m5.128s
user	0m1.829s
sys	0m0.729s

```

Tried it with Satellite 6.8 and it made no difference, hovered around 57s:

```
real	0m57.316s
user	0m1.954s
sys	0m0.800s
```

:+1:

**Comment by mccun934** [comment]
we need to improve the Rails loading time in Foreman, but this is a good patch to include in the mean time

File: definitions/features/service.rb [hunk 1]
```diff
@@ -61,29 +64,27 @@ def use_system_service(action, options, spinner)
...
+    filtered_services(options).each_value do |group|
+      group.each_slice(2).to_a.each do |subgroup|
+        spinner.update("#{action_noun(action)} #{subgroup.join(', ')}")
+      end
+      result = execute_with_status('systemctl ' \
+        "#{action} #{group.map(&:name).join(' ')}")
+      if result[0] > 0 || action == 'status'
+        failed_services |= services_status(action, group, options, spinner)
       end
+      status = result[0] if result[0] > 0
```
Comment (kgaikwad) [inline definitions/features/service.rb:76]:
```suggestion
      systemctl_status, _output = execute_with_status('systemctl ' \
        "#{action} #{group.map(&:name).join(' ')}")
      if systemctl_status > 0 
        status = systemctl_status
        failed_services |= services_status(action, group, options, spinner)
      end
```

File: definitions/features/service.rb [hunk 2]
```diff
@@ -61,29 +64,27 @@ def use_system_service(action, options, spinner)
...
-      if item_status > 0
-        status = item_status
-        failed_services << service
+    filtered_services(options).each_value do |group|
+      group.each_slice(2).to_a.each do |subgroup|
+        spinner.update("#{action_noun(action)} #{subgroup.join(', ')}")
+      end
+      result = execute_with_status('systemctl ' \
+        "#{action} #{group.map(&:name).join(' ')}")
+      if result[0] > 0 || action == 'status'
```
Comment (kgaikwad) [inline definitions/features/service.rb:73]: One query - why did you add special handling for action == `status` here and inside `format_status` method? As per my understanding it is because you clubbed two things - failed_services & formatting the message logic in a single method, right?

I found it little confusing while reading the code, will it be possible to separate out these two things

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Hello @upadhyeammit, 
I haven't tested the changes. Added few inline comments.

File: definitions/features/service.rb [hunk 3]
```diff
@@ -31,7 +30,11 @@ def filtered_services(options)
     service_list = filter_services(service_list, options)
```
Comment (kgaikwad) [inline definitions/features/service.rb:30]: As the return type of `filter_services` is modified from `array` to `hash` so could you rename this `service_list` variable to `services` or something else?

File: definitions/features/service.rb (hunk 2 above)
Comment (upadhyeammit) [reply definitions/features/service.rb:73]: Yes, I agree its bit confusing. Actually as services are being handled in group I need to sequentially get the status of every service if there is failure or action is status. I dont see a way to simply it more than this, at max I can remove the OR condition on line 73 and use elsif because it exactly needs same methods with different params.
With this I better I change failed_services variable to `statuses` as it resembles this implementation.

**Review by kgaikwad — CHANGES_REQUESTED** [review]
Added a small inline comment. 
I haven't tested the changes, code wise rest of the changes look good to me.

**Review by kgaikwad — APPROVED** [review]
LGTM codewise :+1:
